### PR TITLE
MCA/UCX: atomic add/swap are moved to new UCX atomic API

### PR DESCRIPTION
**Pull request by hoopoepg**
Signed-off-by: Sergey Oblomov <sergeyo@mellanox.com>

### CODE REVIEW

**Comment by hoopoepg** [comment]
@yosefe could you look?

File: oshmem/mca/atomic/ucx/atomic_ucx_module.c [hunk 1]
```diff
@@ -49,3 +49,24 @@ mca_atomic_ucx_query(int *priority)
     return NULL ;
 }
 
+void mca_atomic_ucx_complete_cb(void *request, ucs_status_t status)
+{
+}
+
+ucs_status_t mca_atomic_ucx_wait_request(ucs_status_ptr_t request)
```
Comment (hoopoepg) [inline oshmem/mca/atomic/ucx/atomic_ucx_module.c]: make it inline

File: oshmem/mca/atomic/ucx/atomic_ucx_module.c [hunk 2]
```diff
@@ -49,3 +49,24 @@ mca_atomic_ucx_query(int *priority)
...
+
+    /* check for request completed or failed */
+    if (UCS_OK == request) {
+        return UCS_OK;
+    } else if (UCS_PTR_IS_ERR(request)) {
+        return UCS_PTR_STATUS(request);
+    }
+
+    while (UCS_INPROGRESS == (status = ucp_request_check_status(request))) {
+        ucp_worker_progress(mca_spml_self->ucp_worker);
```
Comment (hoopoepg) [inline oshmem/mca/atomic/ucx/atomic_ucx_module.c]: add opal_progress

**Comment by hoopoepg** [comment]
@yosefe is it ok for now?

File: oshmem/mca/atomic/ucx/atomic_ucx_cswap.c [hunk 3]
```diff
@@ -27,23 +27,23 @@ int mca_atomic_ucx_cswap(void *target,
...
     uint64_t rva;
+    uint64_t val;
+
+    if ((8 != nlong) && (4 != nlong)) {
+        ATOMIC_ERROR("[#%d] Type size must be 4 or 8 bytes.", my_pe);
+        return OSHMEM_ERROR;
+    }
 
     ucx_mkey = mca_spml_ucx_get_mkey(pe, target, (void *)&rva); 
+    val = (8 == nlong) ? *(uint64_t*)value : *(uint32_t*)value;
```
Comment (yosefe) [inline oshmem/mca/atomic/ucx/atomic_ucx_cswap.c]: can we unite this check with line 35?
e.g
if (nlong==8) val=.. else if(nlong==4) val=.. else error
(for fadd as well)

File: oshmem/mca/atomic/ucx/atomic_ucx_cswap.c [hunk 4]
```diff
@@ -27,23 +27,23 @@ int mca_atomic_ucx_cswap(void *target,
...
-                break;
-            case 8:
-                status = ucp_atomic_swap64(mca_spml_self->ucp_peers[pe].ucp_conn, 
-                        *(uint64_t *)value, rva, ucx_mkey->rkey, prev);
-                break;
-            default:
-                goto err_size;
-        }
+        status_ptr = ucp_atomic_fetch_nb(mca_spml_self->ucp_peers[pe].ucp_conn,
+                UCP_ATOMIC_FETCH_OP_SWAP, val, prev, nlong,
```
Comment (yosefe) [inline oshmem/mca/atomic/ucx/atomic_ucx_cswap.c]: indentation

File: oshmem/mca/atomic/ucx/atomic_ucx_cswap.c [hunk 5]
```diff
@@ -56,15 +56,13 @@ int mca_atomic_ucx_cswap(void *target,
                         *(uint64_t *)cond, *(uint64_t *)value, rva, ucx_mkey->rkey, prev);
                 break;
             default:
-                goto err_size;
+                assert(0); /* should not be here */
+                status = UCS_ERR_INVALID_PARAM;
```
Comment (yosefe) [inline oshmem/mca/atomic/ucx/atomic_ucx_cswap.c]: need to use NB atomics for cswap as well

File: oshmem/mca/atomic/ucx/atomic_ucx_fadd.c [hunk 6]
```diff
@@ -26,43 +26,29 @@ int mca_atomic_ucx_fadd(void *target,
...
-                break;
-            case 8:
-                status = ucp_atomic_add64(mca_spml_self->ucp_peers[pe].ucp_conn, 
-                        *(uint64_t *)value, rva, ucx_mkey->rkey);
-                break;
-            default:
-                goto err_size;
-        }
+        status = ucp_atomic_post(mca_spml_self->ucp_peers[pe].ucp_conn, 
+                UCP_ATOMIC_POST_OP_ADD, val, nlong, rva, ucx_mkey->rkey);
```
Comment (yosefe) [inline oshmem/mca/atomic/ucx/atomic_ucx_fadd.c]: indentation

File: oshmem/mca/atomic/ucx/atomic_ucx_fadd.c [hunk 7]
```diff
@@ -26,43 +26,29 @@ int mca_atomic_ucx_fadd(void *target,
...
-                break;
-            case 8:
-                status = ucp_atomic_fadd64(mca_spml_self->ucp_peers[pe].ucp_conn, 
-                        *(uint64_t *)value, rva, ucx_mkey->rkey, prev);
-                break;
-            default:
-                goto err_size;
-        }
+        status_ptr = ucp_atomic_fetch_nb(mca_spml_self->ucp_peers[pe].ucp_conn, 
+                UCP_ATOMIC_FETCH_OP_FADD, val, prev, nlong,
```
Comment (yosefe) [inline oshmem/mca/atomic/ucx/atomic_ucx_fadd.c]: indentation

**Comment by hoopoepg** [comment]
@yosefe anything else to fix?

File: oshmem/mca/atomic/ucx/atomic_ucx_cswap.c [hunk 8]
```diff
@@ -27,44 +27,48 @@ int mca_atomic_ucx_cswap(void *target,
...
-                status = ucp_atomic_cswap32(mca_spml_self->ucp_peers[pe].ucp_conn, 
-                        *(uint32_t *)cond, *(uint32_t *)value, rva, ucx_mkey->rkey, prev);
-                break;
-            case 8:
-                status = ucp_atomic_cswap64(mca_spml_self->ucp_peers[pe].ucp_conn, 
-                        *(uint64_t *)cond, *(uint64_t *)value, rva, ucx_mkey->rkey, prev);
-                break;
-            default:
-                goto err_size;
+        if (8 == nlong) {
```
Comment (yosefe) [inline oshmem/mca/atomic/ucx/atomic_ucx_cswap.c]: let's try to branch on "nlong" only once. currently, for cswap, it's done 3 times

Comment (hoopoepg) [reply oshmem/mca/atomic/ucx/atomic_ucx_cswap.c]: ok, only one used

File: oshmem/mca/atomic/ucx/atomic_ucx_cswap.c [hunk 9]
```diff
@@ -27,44 +27,40 @@ int mca_atomic_ucx_cswap(void *target,
...
 {
     ucs_status_t status;
+    ucs_status_ptr_t status_ptr;
     spml_ucx_mkey_t *ucx_mkey;
     uint64_t rva;
+    uint64_t val;
+    uint64_t cmp;
+
+    if (8 == nlong) {
+        val = *(uint64_t*)value;
```
Comment (yosefe) [inline oshmem/mca/atomic/ucx/atomic_ucx_cswap.c]: still 3 branches.. memcpy is also a branch
ok how about this:
```
static inline mca_atomic_ucx_cswap_do(..., int nlong, ...) {
 ... 
}

mca_atomic_ucx_cswap(...) {
 if (8 == nlong) {
   mca_atomic_ucx_cswap_do(..., 8, ...);
 } else if (4 == nlong) {
  mca_atomic_ucx_cswap_do(..., 4, ...);
 } else {
   error
 }
}
```

Comment (hoopoepg) [reply oshmem/mca/atomic/ucx/atomic_ucx_cswap.c]: @yosefe done

**Pull request merged**